src/dir/glob.cr

```diff
@@ -1,266 +1,297 @@
+  def self.glob(*patterns, match_hidden = false, &block : String -> _)
+    glob(patterns, match_hidden: match_hidden) do |path|
+      yield path
+    end
+  end
 
-      recursion_depth = ptrn.count(File::SEPARATOR)
+  # ditto
+  def self.glob(patterns : Enumerable(String), match_hidden = false, &block : String -> _)
+    Globber.new(match_hidden: match_hidden).glob(patterns) do |path|
+      yield path
+    end
+  end
 
-      if starts_with_path_separator
-        dir = root
-      else
-        dir = cwd
+  # :nodoc:
+  struct Globber
+    record DirectoriesOnly
+    record ConstantEntry, path : String
+    record EntryMatch, pattern : String do
+      def matches?(string)
+        File.match?(pattern, string)
       end
-
-      if ptrn.includes? "**"
-        recursion_depth = Int32::MAX
+    end
+    record RecursiveDirectories
+    record ConstantDirectory, path : String
+    record RootDirectory
+    record DirectoryMatch, pattern : String do
+      def matches?(string)
+        File.match?(pattern, string)
       end
+    end
+    alias PatternType = DirectoriesOnly | ConstantEntry | EntryMatch | RecursiveDirectories | ConstantDirectory | RootDirectory | DirectoryMatch
 
-      # optimize the glob by starting with the directory
-      # which is as nested as possible:
-      lastidx = 0
-      depth = 0
-      escaped = false
-      last_is_file_separator = false
-      count = 0
-      nested_path = String.build do |str|
-        ptrn.each_char_with_index do |c, i|
-          if c == '\\'
-            escaped = true
-            last_is_file_separator = false
-            next
-          elsif c == File::SEPARATOR
-            unless last_is_file_separator
-              depth += 1
-              str << c
-              count += 1
-            end
-            lastidx = count
-            last_is_file_separator = true
-          elsif !escaped && special.includes? c
-            break
-          else
-            last_is_file_separator = false
-            str << c
-            count += 1
+    property? match_hidden
+
+    def initialize(@match_hidden = false)
+    end
+
+    def glob(patterns : Enumerable(String), &block : String -> _)
+      patterns.each do |pattern|
+        sequences = compile(pattern)
+
+        sequences.each do |sequence|
+          run(sequence) do |match|
+            yield match
           end
-          escaped = false
         end
       end
-      nested_path = nested_path[0...lastidx]
+    end
+
+    private def compile(pattern)
+      expanded_patterns = [] of String
+      File.expand_brace_pattern(pattern, expanded_patterns)
 
-      recursion_depth -= depth if recursion_depth != Int32::MAX
-      dir = File.join(dir, nested_path)
-      if !nested_path.empty? && nested_path[0] == File::SEPARATOR
-        nested_path = nested_path[1..-1]
+      expanded_patterns.map do |expanded_pattern|
+        single_compile expanded_pattern
       end
+    end
+
+    private def single_compile(glob)
```

Comment:
Why not name this `compile` and make it an overload?

Reply:
It's not a public API and only called once in the `compile` method. It could even be directly embedded but I think it's more clear to have a separate method.
But what benefit would it have to name both methods the same? Why add unneccessary work for the type inference? =)

Reply:
True - it's a style issue. I dislike the compiler performance appeal though, it's not an issue. The benefit to me is it looks nicer.